src/librustc_mir/borrow_check/mod.rs

```diff
@@ -1808,15 +1724,162 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
+                                                elem: ProjectionElem::Deref })) => {
+                let local_decl = &self.mir.local_decls[*local];
+                let locations = self.mir.find_assignments(*local);
+                let (err_help_span, suggested_code) = 'go: loop {
+                    if locations.len() > 0 {
+                        let assignment_rhs_span = self.mir.source_info(locations[0]).span;
+                        let snippet = self.tcx.sess.codemap().span_to_snippet(assignment_rhs_span);
+                        if let Ok(src) = snippet {
+                            let borrowed_expr = src[1..].to_string();
+                            break 'go (assignment_rhs_span, format!("&mut {}", borrowed_expr));
+                        }
+                    }
+                    let ty_mut = local_decl.ty.builtin_deref(true).unwrap();
+                    assert_eq!(ty_mut.mutbl, hir::MutImmutable);
+                    break 'go (local_decl.source_info.span, format!("&mut {}", ty_mut.ty));
+                };
+                err.span_suggestion(err_help_span,
+                                    "consider changing this to be a mutable reference",
+                                    suggested_code);
+
+                if let Some(name) = local_decl.name {
+                    err.span_label(span,
+                                   format!("`{}` is a `&` reference, \
+                                            so the data it refers to cannot be written", name));
+                } else {
+                    err.span_label(span, "cannot assign through `&`-reference");
+                }
+                note_the_place_err = false;
+            }
+
+            (ErrorAccessKind::MutableBorrow,
+             Place::Projection(box Projection { base: Place::Local(local),
+                                                elem: ProjectionElem::Deref }))
+                if self.mir.local_decls[*local].is_user_variable.is_some() =>
+            {
+                let local_decl = &self.mir.local_decls[*local];
+                let err_help_span = local_decl.source_info.span;
+                let ty_mut = local_decl.ty.builtin_deref(true).unwrap();
+                assert_eq!(ty_mut.mutbl, hir::MutImmutable);
+                let suggested_code = format!("&mut {}", ty_mut.ty);
+                err.span_suggestion(err_help_span,
+                                    "consider changing this to be a mutable reference",
+                                    suggested_code);
+
+                note_the_place_err = true;
+            }
+
+            (ErrorAccessKind::MutableBorrow, _) => {
+                note_the_place_err = true
+            }
+
+            (ErrorAccessKind::Mutate, _) => {
+                err.span_label(span, "cannot mutate");
+                note_the_place_err = true;
+            }
+        }
+
+        if note_the_place_err && place != the_place_err {
```

Comment:
I personally would remove this message. If we can identify such a place, we should suggest how to change it to mutable. Otherwise, I suspect it will confuse users as much as anything.

Reply:
My plan is indeed to remove this message once we’ve added all the diagnostic cases of interest